plutus-ir/src/Language/PlutusIR/Rename.hs

```diff
@@ -0,0 +1,131 @@
+-- instance in scope.
+module Language.PlutusIR.Rename () where
+
+import           Language.PlutusIR
+
+import qualified Language.PlutusCore                 as PLC
+import qualified Language.PlutusCore.Name            as PLC
+import qualified Language.PlutusCore.Rename          as PLC
+import qualified Language.PlutusCore.Rename.Internal as PLC
+
+import           Control.Monad.Reader
+import           Control.Monad.Trans.Class           (lift)
+import           Control.Monad.Trans.Cont            (ContT (..))
+
+{- Note [Renaming of mutually recursive bindings]
+The 'RenameM' monad is a newtype wrapper around @ReaderT renaming Quote@, so in order to bring
+a name into the scope we need to use the following pattern:
+
+    withFreshenedX :: X -> (X -> RenameM renaming c) -> RenameM renaming c
+
+where an 'X' is some value that has names inside and the second argument is a continuation that
+expects to receive this value, but with all the names freshened and with the mappings from the
+old uniques to the new ones saved. For the latter reason we can't just write
+
+    freshenX :: X -> RenameM renaming X
+
+because if you write
+
+    freshenX x >>= \xFr -> ...
+
+then the mappings from the uniques of 'x' to the uniques of 'xFr' do not exist in @...@.
+
+Two problems arise:
+
+1. If you have a list of @X@s and then an @Y@ and then another list of @Z@s and you need to freshen
+   all of those, it's not nice to compose @withFreshened*@ functions at all. In order to make these
+   functions composable, we can wrap them in the 'ContT' monad. See e.g. 'renameDatatypeCM' below.
+2. Handling of mutually recursive bindings requires to bring a bunch of bindings in scope first and
+   only then to rename their RHSs. I.e. one part of a 'ScopedRenameM' computation must be performed
+   immediately (brinding names in scope) and the other part must be postponed until all bindings
+   are in scope (renaming of RHSs). Hence we have the following pattern:
+
+       renameXCM :: X -> ContT c PLC.ScopedRenameM (PLC.ScopedRenameM X)
+
+   where an 'X' is some value that defines some names and has types/terms that need to be renamed.
+   The first 'PLC.ScopedRenameM' is for bringing names in scope and the second 'PLC.ScopedRenameM'
+   is for performing the renaming.
+-}
+
+instance (PLC.HasUnique (tyname ann) PLC.TypeUnique, PLC.HasUnique (name ann) PLC.TermUnique) =>
+        PLC.Rename (Term tyname name ann) where
+    rename = PLC.runScopedRenameM . renameTermM
+
+-- | Rename a 'Datatype' in the CPS-transformed 'ScopedRenameM' monad.
+renameDatatypeCM
+    :: (PLC.HasUnique (tyname ann) PLC.TypeUnique, PLC.HasUnique (name ann) PLC.TermUnique)
+    => Datatype tyname name ann
+    -> ContT c PLC.ScopedRenameM (PLC.ScopedRenameM (Datatype tyname name ann))
+renameDatatypeCM (Datatype x dataDecl params matchName constrs) = do
+    dataDeclFr <- ContT $ PLC.withFreshenedTyVarDecl dataDecl
```

Comment:
> Given that `withFreshenedTyVarDecl` is already sort of a continuation passing thing, would it make sense to put it in `Cont` to begin with?

I'm just being consistent: `withFreshenedTyVarDecl` is very similar to `withFreshenedName` and the latter is more useful without `ContT`. We could have a separate family of combinators that are `ContT`-wrapped, but that would clutter the API. It's really just easier to manually wrap in `ContT` when needed.

Reply:
I guess I'm really saying: why not put the PLC renamer in `ContT` too?

Reply:
> `Reader` automatically manages scope for us and that's all we need.

That's the main thing here. With `State` we would just add more and more bound variables and never delete them from the state even when they become out of scope. That's not the case with `Reader` where environment extensions closely follow the scope structure, so that when a variable becomes out of scope, it's no longer in the environment as well, because the effect of `withFreshenedName` is local:

```haskell
withFreshenedName name k = do
    uniqNew <- coerce <$> freshUnique
    local (insertByNameM name uniqNew) $ k (name & unique .~ uniqNew)
```

Reply:
:open_mouth: :boom: 

I don't entirely understand how the addition of `Cont` makes the accumulation of renamings non-local. I can see that we *could* use `Cont` cleverness to move our bindings around, but afaict we're not doing that, and we still end up running nested `Reader` computations with appropriate scope.

I don't think it's crazy to shove `Cont` in there even if we're only using it for a few functions (in this case the refreshing functions). Surely it shouldn't affect anything else if we don't use it there?

Reply:
> Surely it shouldn't affect anything else if we don't use it there?

It does affect things. [Here](https://gist.github.com/effectfully/5dd99d02a24a1b5e7792b08f91fcbb35) is a renamer from `language-plutus-core` that uses `ContT`. Tests do not pass (although I see errors that I cannot explain, but I'm really glad I see them, because they puzzled me when I saw them being throwed from the annotation machinery which was wrongly implemented in terms of `State`). Maybe you can change something in my code, but I really doubt you can make it work.

What `ContT` allows to do and what I use it for is collecting bindings, like if you have

```haskell
let x = ...
    y = ...
    z = ...
```

then instead of writing

```haskell
withFreshenedX x $ \xFr ->
    withFreshenedY y $ \yFr -> 
        withFreshenedZ z $ \zFr -> 
            k xFr yFr zFr
```

(which is really awkward), you can write

```haskell
runContT (do
    xFr <- ContT $ withFreshenedX x
    yFr <- ContT $ withFreshenedY y
    zFr <- ContT $ withFreshenedZ z
    pure (xFr, yFr, zFr)
    ) $ \(xFr, yFr, zFr) -> k xFr yFr zFr
```

and use `traverse` when lists are involved, etc. But once you bring all the variables in scope, you have to leave `ContT`, because otherwise the whole computation will be of this form:

```haskell
withFreshenedX x1 $ \x1Fr ->
    withFreshenedX x2 $ \x2Fr ->
        withFreshenedX x3 $ \x3Fr ->
            withFreshenedX x4 $ \x4Fr ->
                ...
```

i.e. we'll always nest one `with` inside another, while we do want most of the time not to nest.

So `ContT` is specifically for dealing with nested computations of this form:

`a -> (a -> RenameM renaming c) -> RenameM renaming c`

in a nice monadic way. But when nesting is not desired (and most of the time it's not desired), then `ContT` shouldn't be used, because the semantics of

```haskell
do
    y1 <- withFreshenedX x1 $ \x1Fr -> k1 x1Fr
    y2 <- withFreshenedX x2 $ \x2Fr -> k2 x2Fr
    ...
```

is very different from the semantics of

```haskell
do
    x1Fr <- ContT $ withFreshenedX x1
    y1 <- k1 x1Fr
    x2Fr <- ContT $ withFreshenedX x2
    y2 <- k2 x2Fr
    ...
```

(the types of `k1` and `k2` now mention `ContT` unlike in the previous snippet) because the latter elaborates to

```haskell
withFreshenedX x1 $ \x1Fr ->
    k1 x1Fr $ \y1 ->
        withFreshenedX x2 $ \x2Fr ->
            k2 x2Fr $ \y2 ->
                ...
```

And that is how `Reader` becomes `State`.

You can see the nesting in the definition of the `Monad` instance for `ContT`:

```haskell
m >>= k  = ContT $ \ c -> runContT m (\ x -> runContT (k x) c)
```

Now substitute `ContT $ withA A` for `m` and you'll get

```
... runContT (ContT $ withA a) (\x -> <the_rest_of_the_computation>)
```

which simplifies to

```
... withA a (\x -> <the_rest_of_the_computation>)
```

Reply:
Ugh, `Cont` always blows my mind. I believe you, but it still seems annoyingly fiddly.